twistes/bulk_utils.py

```diff
@@ -0,0 +1,197 @@
+import json
+from operator import methodcaller
+
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from twistes.compatability import string_types
+from twistes.consts import EsBulk, EsDocProperties
+from twistes.exceptions import BulkIndexError, ConnectionTimeout
+
+
+class ActionParser(object):
+    @staticmethod
+    def expand_action(data):
+        """
+        From one document or action definition passed in by the user extract the
+        action/data lines needed for elasticsearch's
+        :meth:`~elasticsearch.Elasticsearch.bulk` api.
+        """
+        # when given a string, assume user wants to index raw json
+        if isinstance(data, string_types):
+            return '{"index": {}}', data
+
+        # make sure we don't alter the action
+        data = data.copy()
+        op_type = data.pop(EsBulk.OP_TYPE, EsBulk.INDEX)
+        action = {op_type: {}}
+        for key in (EsDocProperties.INDEX, EsDocProperties.PARENT, EsDocProperties.PRECOLATE,
```

Comment:
extract to method - get relevant fields, put the tuple in a variable with satisfying name